### PR TITLE
Gradle build fixes to support Java 17 and Gradle 7.3

### PR DESCRIPTION
**Pull request by pepone**
This PR updates Java build to work with Java 17 and Gradle 7.3

This requires a new version of the slice Gradle plug-in, for testing, you can use the one from https://github.com/pepone/ice-builder-gradle/tree/gradle-7.x I will open a separate PR once I sort out the maven stuff there.

### CODE REVIEW

File: java/build.gradle [hunk 1]
```diff
@@ -37,6 +37,11 @@ subprojects {
         mavenCentral()
     }
 
+    // This setting is required for Java modules when target Java 8
```
[the remark this answers is not in the document]
Comment (externl) [reply java/build.gradle:40]: Should we have a special Java 8 check?

File: java/gradle/library.gradle [hunk 2]
```diff
@@ -85,44 +31,20 @@ task javadocJar(type: Jar, dependsOn: javadoc) {
...
-        ( project.hasProperty('signing.keyId') &&
-          project.hasProperty('signing.password') &&
-          project.hasProperty('signing.secretKeyRingFile') )
-    }
-    sign configurations.archives
+// The maven plugin change in Gradle 7.3, for compatibility with older gradle versions
+// we split maven related tasks to separate files that use the corresponding builtin gradle
+// plug-ins
+project.ext.pomName = "${libDir}/${project.name}-${project.version}.pom"
+if(GradleVersion.current() >= GradleVersion.version('7.3')) {
```
Comment (externl) [inline java/gradle/library.gradle:38]: Aren't we always using the local Gradle version we have specified? Or is this really more for the offline Debian builds?

If so I would just skip the publishing stuff if we can. We only need to publish with one version I think.

File: java/gradle/library.gradle [hunk 3]
```diff
@@ -85,44 +31,20 @@ task javadocJar(type: Jar, dependsOn: javadoc) {
...
+    apply from: "$project.ext.topSrcDir/java/gradle/maven-publish.publishing.gradle"
+} else {
+    apply from: "$project.ext.topSrcDir/java/gradle/maven.publishing.gradle"
 }
 
-ext.mavenRepo = rootProject.hasProperty('mavenRepo')?project.mavenRepo:System.getenv('MAVEN_REPO')?:''
-ext.mavenSnapshotRepo = rootProject.hasProperty('mavenSnapshotRepo')?project.mavenSnapshotRepo:System.getenv('MAVEN_SNAPSHOT_REPO')?:''
-ext.mavenUser = rootProject.hasProperty('mavenUser')?project.mavenUser:System.getenv('MAVEN_USER')?:''
-ext.mavenPassword = rootProject.hasProperty('mavenPassword')?project.mavenPassword:System.getenv('MAVEN_PASSWORD')?:''
```
Comment (externl) [inline java/gradle/library.gradle:106]: Do we not need this configuration settings at all now?

Comment (pepone) [reply java/gradle/library.gradle:106]: AFAIK we don't use that for our maven upload.  We have a separate dist-util job for uploading maven artifacts.

File: java/gradle/library.gradle (hunk 2 above)
Comment (pepone) [reply java/gradle/library.gradle:38]: Right that is for Debian packaging that needs to run offline and with the local Gradle executable.

File: java/gradle/library.gradle (hunk 3 above)
Comment (externl) [reply java/gradle/library.gradle:106]: Ok, we used to use this, but not anymore?

File: java/build.gradle (hunk 1 above)
[the remark this answers is not in the document]
Comment (pepone) [reply java/build.gradle:40]: we can add one yes

**Review by externl — APPROVED** [review]

**Pull request merged**